### PR TITLE
chore(auth): fix AWSDataStorePluginAuthCognitoTests to build 

### PR DESCRIPTION
**Pull request by manaswi223**
*Issue #, if available:*

*Description of changes:*

*Check points: (check or cross out if not relevant)*

- [ ] Added new tests to cover change, if needed
- [ ] Build succeeds with all target using Swift Package Manager
- [ ] All unit tests pass
- [ ] All integration tests pass
- [ ] Security oriented best practices and standards are followed (e.g. using input sanitization, principle of least privilege, etc)
- [ ] Documentation update for the change if required
- [ ] PR title conforms to conventional commit style
- [ ] If breaking change, documentation/changelog update with migration instructions

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift [hunk 1]
```diff
@@ -353,24 +309,24 @@ extension AWSDataStoreAuthBaseTest {
...
+                onFailure(error as! DataStoreError)
             }
         }
-        receiveValue: { posts in
-            XCTAssertNotNil(posts)
-            expectations.query.fulfill()
-        }.store(in: &requests)
-        wait(for: [expectations.query],
-             timeout: 60)
+        await waitForExpectations([expectations.query], timeout: 60)
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift:322]: this looks good! The task is created, so code waits on line 322, then when the awaited call in the task finishes (and does not hang) then `expectations.query` is fulfilled, causing line 322 to continue

File: AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift [hunk 2]
```diff
@@ -379,40 +335,42 @@ extension AWSDataStoreAuthBaseTest {
...
+        await waitForExpectations([expectations.subscriptionsEstablished,
+                                   expectations.modelsSynced,
+                                   expectations.ready],
+                             timeout: 60)
+    }
+    
+    enum crudError: Error {
+        case notSaved(Error: String?)
+        case notDeleted(Error: String?)
     }
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift:365]: what's this?

File: AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift [hunk 3]
```diff
@@ -424,40 +382,31 @@ extension AWSDataStoreAuthBaseTest {
...
+            do {
+                let posts = try await Amplify.DataStore.save(model)
+                XCTAssertNotNil(posts)
+                let deletedposts: () = try await Amplify.DataStore.delete(posts)
+                XCTAssertNotNil(deletedposts)
+            } catch crudError.notDeleted(let error) {
+                XCTFail("Unable to delete posts from Datastore with errors: \(String(describing: error))")
+            } catch crudError.notSaved(let error) {
+                XCTFail("Unable to save posts from Datastore with errors: \(String(describing: error))")
             }
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift:405]: DataStore.save and DataStore.delete both throw DataStore error, so you can catch it like this:
```
} catch let error as DataStoreError {
```

File: AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreCategoryPluginAuthIntegrationTests.swift [hunk 4]
```diff
@@ -39,7 +39,7 @@ class AWSDataStoreCategoryPluginAuthIntegrationTests: AWSDataStoreAuthBaseTest {
             return
         }
 
-        let syncReceivedInvoked = expectation(description: "Received SyncReceived event")
+//        let syncReceivedInvoked = expectation(description: "Received SyncReceived event")
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreCategoryPluginAuthIntegrationTests.swift:42]: `        let syncReceivedInvoked = asyncExpectation(description: "Received SyncReceived event")` ?

File: AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreCategoryPluginAuthIntegrationTests.swift [hunk 5]
```diff
@@ -49,17 +49,17 @@ class AWSDataStoreCategoryPluginAuthIntegrationTests: AWSDataStoreAuthBaseTest {
...
         guard try await HubListenerTestUtilities.waitForListener(with: syncReceivedListener, timeout: 5.0) else {
             XCTFail("syncReceivedListener registered for hub")
             return
         }
 
-        signIn(user: user1)
+        try await signIn(user: user1)
 
-        wait(for: [syncReceivedInvoked], timeout: TestCommonConstants.networkTimeout)
+//        wait(for: [syncReceivedInvoked], timeout: TestCommonConstants.networkTimeout)
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreCategoryPluginAuthIntegrationTests.swift:62]: `await waitFor...`

**Comment by lawmicha** [comment]
Instead of reopening another PR because you've create a new branch like `manaswi.datastore-cognitoAuth`. You can take the previous branch `manaswi.datastore-cognito` from the previous PR https://github.com/aws-amplify/amplify-ios/pull/2251 and delete it locally:
```
git branch -D manaswi.datastore-cognito
```

Then rename `manaswi.datastore-cognitoAuth` to `manaswi.datastore-cognito` 
```
// assuming you are on `manaswi.datastore-cognitoAuth` branch
git branch -M manaswi.datastore-cognito
```

Then you can do a force push to `manaswi.datastore-cognito` and overwrite the commit in the previous PR
```
git push -u origin HEAD --force
```

File: AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift [hunk 6]
```diff
@@ -325,21 +289,13 @@ extension AWSDataStoreAuthBaseTest {
...
-        Amplify.DataStore.query(M.self, byId: id) { result in
-            switch result {
-            case .success(let model):
-                queriedModel = model
-                queriedInvoked.fulfill()
-            case .failure(let error):
-                XCTFail("Failed to query model \(error)", file: file, line: line)
-            }
+        do {
+            queriedModel = try await Amplify.DataStore.query(M.self, byId: id)
```
[the remark this answers is not in the document]
Comment (manaswi223) [reply AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift:295]: ya, I will change back to this (wrote this i previous PR, changed after you suggested)

File: AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift [hunk 7]
```diff
@@ -75,27 +76,27 @@ class AWSDataStoreAuthBaseTest: XCTestCase {
...
+        try await signOut()
         await Amplify.reset()
     }
 
     // MARK: - Test Helpers
     func makeExpectations() -> AuthTestExpectations {
         AuthTestExpectations(
-            subscriptionsEstablished: expectation(description: "Subscriptions established"),
-            modelsSynced: expectation(description: "Models synced"),
+            subscriptionsEstablished: AsyncExpectation(description: "Subscriptions established"),
```
[the remark this answers is not in the document]
Comment (manaswi223) [reply AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift:88]: sure, changed `AsyncExpectation` -> `asyncExpectation` . Didn't know this existed :)

File: AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift [hunk 8]
```diff
@@ -203,117 +197,87 @@ extension AWSDataStoreAuthBaseTest {
...
-                signInInvoked.fulfill() // won't count as pass
-            case .success:
-                signInInvoked.fulfill()
-            }
+        do {
+            let response = try await Amplify.Auth.signIn(username: user.username,
+                                                     password: user.password, options: nil)
+            print("received response for signIn \(response)")
+            let result = try await isSignedIn()
+            XCTAssert(result)
```
[the remark this answers is not in the document]
Comment (manaswi223) [reply AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift:210]: sure, changed the naming

File: AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift [hunk 9]
```diff
@@ -203,117 +197,87 @@ extension AWSDataStoreAuthBaseTest {
...
-
-            case .success:
-                signoutInvoked.fulfill()
-            }
+                 line: UInt = #line) async throws {
+        do {
+            _ = try await Amplify.Auth.signOut()
+            print("signOut successfull")
+            let result = try await !isSignedIn()
+            XCTAssert(result)
```
[the remark this answers is not in the document]
Comment (manaswi223) [reply AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift:226]: done !

File: AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift [hunk 10]
```diff
@@ -203,117 +197,87 @@ extension AWSDataStoreAuthBaseTest {
...
+            //XCTAssert(!isSignedIn(), file: file, line: line)
+            //rhs: @autoclosure () async throws -> Bool
+            //lhs: @autoclosure () throws -> Bool
+        } catch {
+            print("Failed to signOut user with error: \(error)")
```
[the remark this answers is not in the document]
Comment (manaswi223) [reply AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift:231]: done

File: AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift [hunk 11]
```diff
@@ -203,117 +197,87 @@ extension AWSDataStoreAuthBaseTest {
...
             return ""
         }
-
         return result
     }
 
-    func getIdentityId() -> String {
-        let retrieveIdentityCompleted = expectation(description: "retrieve identity completed")
+    func getIdentityId() async throws -> String! {
         var resultOptional: String?
```
[the remark this answers is not in the document]
Comment (manaswi223) [reply AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift:270]: Michael, in this code:
we are using resultOptional to check the final result  if we actually got the queried result or not. (this is my understanding)
but if we just return userSub, the result if the query was successful/not case cannot be inferred? what are your thoughts?

File: AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift (hunk 2 above)
Comment (manaswi223) [reply AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift:356]: removed it: thought of catching error-specific failures in the catch of the do-catch block. got clarified with 
`} catch let error as DataStoreError {`

File: AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift [hunk 12]
```diff
@@ -203,117 +197,87 @@ extension AWSDataStoreAuthBaseTest {
...
-        let retrieveUserSubCompleted = expectation(description: "retrieve userSub completed")
+    
+    // trying to catch specific errors to getUserSub() below
+    //    enum AuthError: Error {
+    //        case noAuthSession
+    //        case noAuthSessioncognito
+    //        case noUserSub
+    //    }
+    func getUserSub() async throws -> String! {
         var resultOptional: String?
```
[the remark this answers is not in the document]
Comment (manaswi223) [reply AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift:249]: Michael, in this code:
we are using resultOptional to check the final result if we actually got the queried result or not. (this is my understanding)
but if we just return userSub, the result if the query was successful/not case cannot be inferred? what are your thoughts?

File: AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift [hunk 13]
```diff
@@ -203,144 +197,75 @@ extension AWSDataStoreAuthBaseTest {
...
-                signInInvoked.fulfill()
-            }
+        do {
+            let response = try await Amplify.Auth.signIn(username: user.username,
+                                                     password: user.password, options: nil)
+            print("received response for signIn \(response)")
+            let isSignedIn = try await isSignedIn()
+            XCTAssert(isSignedIn)
+        } catch {
+            print("Failed to signIn user with error: \(error)")
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift:212]:
```suggestion
            XCTFail("Signin failure \(error)", file: file, line: line)
            throw error
```

File: AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift [hunk 14]
```diff
@@ -203,144 +197,75 @@ extension AWSDataStoreAuthBaseTest {
...
-            XCTFail("Could not get isSignedIn for user")
-            return false
+                 line: UInt = #line) async throws {
+        do {
+            _ = try await Amplify.Auth.signOut()
+            print("signOut successfull")
+            let isSignedIn = try await isSignedIn()
+            XCTAssertFalse(isSignedIn)
+        } catch {
+            XCTFail("Failed to signOut user with error: \(error)")
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift:225]:
```suggestion
            XCTFail("SignOut failure \(error)", file: file, line: line)
            throw error
```

File: AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift [hunk 15]
```diff
@@ -203,144 +197,75 @@ extension AWSDataStoreAuthBaseTest {
...
-        wait(for: [retrieveUserSubCompleted], timeout: TestCommonConstants.networkTimeout)
-        guard let result = resultOptional else {
-            XCTFail("Could not get userSub for user")
-            return ""
+    func isSignedIn() async throws -> Bool {
+        do {
+            let authSession = try await Amplify.Auth.fetchAuthSession()
+            return authSession.isSignedIn
+        } catch {
+            fatalError("Failed to get auth session of signed In user with error: \(error)")
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift:234]:
```suggestion
            XCTFail("fetchAuthSession failure \(error)")
            throw error
```

File: AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift [hunk 16]
```diff
@@ -203,144 +197,75 @@ extension AWSDataStoreAuthBaseTest {
...
-                case .success(let identityId):
-                    resultOptional = identityId
-                    retrieveIdentityCompleted.fulfill()
-                case .failure(let error):
-                    XCTFail("Failed to get auth session \(error)")
-                }
-            case .failure(let error):
-                XCTFail("Failed to get auth session \(error)")
+    
+   func getUserSub() async throws -> String {
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift:238]: indentation looks a bit off, maybe ctrl+i this method

File: AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift [hunk 17]
```diff
@@ -203,144 +197,75 @@ extension AWSDataStoreAuthBaseTest {
...
+       do {
+           let authSession = try await Amplify.Auth.fetchAuthSession(options: nil)
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift:240]: i think `options` has a default `nil` value, like above, so no need to pass `nil` in
```suggestion
           let authSession = try await Amplify.Auth.fetchAuthSession()
```

File: AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift [hunk 18]
```diff
@@ -203,144 +197,75 @@ extension AWSDataStoreAuthBaseTest {
...
+           let cognitoAuthSession = authSession as? AuthCognitoIdentityProvider
+           guard let userSub = try cognitoAuthSession?.getUserSub().get() else {
+               throw "Failed to get userSub"
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift:243]: let's fail the test using XCTFail here, i think from our debugging session, it seems a lot easier to pinpoint the problem when it hits XCTFail
```suggestion
               XCTFail("Missing userSub() from session")
               throw "Failed to get userSub"
```

File: AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift [hunk 19]
```diff
@@ -203,144 +197,75 @@ extension AWSDataStoreAuthBaseTest {
...
+           }
+           return userSub
+       } catch {
+           throw error
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift:247]:
```suggestion
           XCTFail("Failed to get auth session \(error)")
           throw error
```

File: AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift [hunk 20]
```diff
@@ -203,144 +197,75 @@ extension AWSDataStoreAuthBaseTest {
...
+                throw "Failed to get identityId"
             }
-        })
-        wait(for: [retrieveIdentityCompleted], timeout: TestCommonConstants.networkTimeout)
-        guard let result = resultOptional else {
-            XCTFail("Could not get identityId for user")
-            return ""
+            return identityId
+        } catch {
+            throw error
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift:260]:
```suggestion
            XCTFail("Failed to get auth session \(error)")
            throw error
```

File: AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift [hunk 21]
```diff
@@ -203,144 +197,75 @@ extension AWSDataStoreAuthBaseTest {
...
+           throw error
+       }
+   }
+    
+    func getIdentityId() async throws -> String! {
+        do {
+            let authSession = try await Amplify.Auth.fetchAuthSession(options: nil)
+            let cognitoAuthSession = authSession as? AuthCognitoIdentityProvider
+            guard let identityId = try cognitoAuthSession?.getIdentityId().get() else {
+                throw "Failed to get identityId"
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift:256]:
```suggestion
                XCTFail("Failed to get identityId")
                throw "Failed to get identityId"
```

File: AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift [hunk 22]
```diff
@@ -379,40 +303,38 @@ extension AWSDataStoreAuthBaseTest {
...
-                    expectations.subscriptionsEstablished.fulfill()
+                    Task { await expectations.subscriptionsEstablished.fulfill() }
                 }
-
                 // modelsSynced fulfilled
                 if event.eventName == dataStoreEvents.modelSynced {
                     modelSyncedCount += 1
                     if modelSyncedCount == expectedModelSynced {
-                        expectations.modelsSynced.fulfill()
+                    Task { await expectations.modelsSynced.fulfill() }
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift:312]: indentation

File: AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift [hunk 23]
```diff
@@ -379,40 +303,38 @@ extension AWSDataStoreAuthBaseTest {
...
-                   expectations.modelsSynced,
-                   expectations.ready],
-             timeout: 60)
-
+        try await Amplify.DataStore.start()
+        
+        await waitForExpectations([expectations.subscriptionsEstablished,
+                                   expectations.modelsSynced,
+                                   expectations.ready],
+                             timeout: 60)
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreAuthBaseTest.swift:327]: indentation

File: AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreCategoryPluginAuthIntegrationTests+Support.swift [hunk 24]
```diff
@@ -11,16 +11,12 @@ import Amplify
...
-            case .failure(let error):
-                XCTFail("Failed to save model \(error)")
-            }
+    func saveModel<T: Model>(_ model: T) async throws {
+        do {
+            let savedposts = try await Amplify.DataStore.save(model)
+            print("Local model was saved: \(savedposts)")
+        } catch {
+            XCTFail("Failed to save model \(error)")
         }
```
Comment (lawmicha) [inline AmplifyPlugins/DataStore/Tests/DataStoreHostApp/AWSDataStorePluginAuthCognitoTests/AWSDataStoreCategoryPluginAuthIntegrationTests+Support.swift:20]: let's apply the single expectation pattern here as well
```suggestion
        let localSaveInvoked = asyncExpectation(description: "local model was saved")
        Task {
            do {
                let savedposts = try await Amplify.DataStore.save(model)
                print("Local model was saved: \(savedposts)")
                await localSaveInvoked.fulfill()
            } catch {
                XCTFail("Failed to save model \(error)")
                throw error
            }
        }
        await waitForExpectations([localSaveInvoked], timeout: TestCommonConstants.networkTimeout)
```

**Review by lawmicha — APPROVED** [review]

**Pull request merged**